### ISSUE TITLE
Add support for SuggestedFixes

### ISSUE THREAD

**Issue by ldez** [golangci/golangci-lint#1779]
**Is your feature request related to a problem? Please describe.**

Add support for SuggestedFixes.

https://github.com/golang/tools/blob/b4639ccb830b1c9602f1c96eba624bbbe20650ea/go/analysis/doc/suggested_fixes.md

**Describe the solution you'd like**

Be able to apply automatically the fixes via `--fix`.

**Describe alternatives you've considered**

Display the suggestion as extra information.

**Additional context**

```go
type Diagnostic struct {
	// ...

	// SuggestedFixes contains suggested fixes for a diagnostic which can be used to perform
	// edits to a file that address the diagnostic.
	// TODO(matloob): Should multiple SuggestedFixes be allowed for a diagnostic?
	// Diagnostics should not contain SuggestedFixes that overlap.
	// Experimental: This API is experimental and may change in the future.
	SuggestedFixes []SuggestedFix // optional


	// ...
}
```
https://github.com/golang/tools/blob/b4639ccb830b1c9602f1c96eba624bbbe20650ea/go/analysis/diagnostic.go#L23-L28

Current linters that support `SuggestedFixes`:
- `goerr113`
- `ruleguard`
- `staticcheck`
- `nlreturn`
- `exportloopref`
- `exhaustive`
- `govet`

Related to #1728

If you want to help me:
- https://bento.me/golangci-lint
- https://bento.me/ldez

Latest update: https://github.com/golangci/golangci-lint/issues/1779#issuecomment-2545783471

**Comment by ldez** [comment]
I'm working on and for now, I'm able to apply suggestions from `goerr113` because the suggestion is only on one line.

For the other, the problem is a bit more complex because the fix must be applied on multiple lines.

**Comment by SVilgelm** [comment]
I think we need to print the suggested fixes if they are not applied (without `--fix`)

**Comment by ldez** [comment]
In some cases, it will br very weird to display the suggestions, because a suggestion can be a `\n` and `TextEdits` is a slice.

**Comment by SVilgelm** [comment]
we can print a `diff`, anyway you are going to create a diff, so we can print replacement, maybe in verbose output

**Comment by SVilgelm** [comment]
@ldez 
I think we need to replace the `Replacement` with `SuggestedFixes` in `golangci-lint`, to avoid a transformation.

**Comment by ldez** [comment]
I think we can support both as the first step, but I'm thinking about I need more time.

**Comment by SVilgelm** [comment]
Here is the code of how to apply SuggestedFixes: https://github.com/golang/tools/blob/54dc8c5edb561003330ec7e649333181bbd0a9bd/go/analysis/internal/checker/checker.go#L316

**Comment by nightlyone** [comment]
@ldez I guess this will be the next major feature after the generics issues are fixed? The developer productivity and happiness enabled by just clicking Apply Suggested fixes is quite significant on a Go ecosystem scale.

**Comment by ldez** [comment]
This issue is not related to GitHub review suggestions.
It's related to the CLI flag `--fix` of the command line.

**Comment by pohly** [comment]
Is there anything that can be done about this on a per-linter basis? In https://github.com/nunnatsa/ginkgolinter/issues/63#issuecomment-1439726442 it was pointed out that for some linters, fixing already works.

**Comment by ldez** [comment]
I planned to handle the topic by a breaking change, I don't want to introduce several ways to handle fixes.

**Comment by bombsimon** [comment]
Would you discourage and/or reject a PR for a linter specific implementation converting suggested fixes to replacements? I've added support for suggested fixers in a linter that only adds or removes a newline so converting it would be fairly simple. However maybe it's better to spend the time on a generic solution instead?

**Comment by ldez** [comment]
I think that it's the right thing for a linter to support `SuggestedFixes` (even if `SuggestedFixes` is still [an experimental API](https://github.com/golang/tools/blob/fbb25cbbd5ff5845639a3035f1fb7ddfd0ac0624/go/analysis/diagnostic.go#L49)), I also believe that we have to remove our custom implementation.

So for now, I think it's better to not introduce partial or linter-specific support/converter.

**Comment by Antonboom** [comment]
`SuggestedFix` is not experimental more.
https://github.com/golang/tools/blob/fa12f34b4218307705bf0365ab7df7c119b3653a/go/analysis/diagnostic.go#L54

**Comment by ldez** [comment]
the commit of the change: https://github.com/golang/tools/commit/dbd600188431770bd182cfa975efa023bea79af1

**Comment by Antonboom** [comment]
Related discussions:

1. https://github.com/golangci/golangci-lint/discussions/2346
2. https://github.com/golangci/golangci-lint/discussions/2298#discussioncomment-1633335

**Comment by StevenACoffman** [comment]
Additional linters that support `SuggestedFixes`:

- `gocritic`
- `whitespace`
- `mirror`
- `misspell`
- `perfsprint`
- `testifylint`
- `usestdlibvars`
- `wsl`

It's not clear to me if `gci` and `gofumpt` run as linters by golangci-lint fully support the SuggestedFixes mechanism, but they can correct their own identified issues.

FYI, until SuggestedFixes for these linters is supported in golangci-lint, you can individually install linters that support SuggestedFixes and run:
```
#!/bin/bash

# this will run all the lint fixes for all the module locally

function is_bin_in_path {
  builtin type -P "$1" &> /dev/null
}

function lintall {
    export GOBIN="$HOME/go/bin"
    ! is_bin_in_path gocritic && go install -v github.com/go-critic/go-critic/cmd/gocritic@latest
    ! is_bin_in_path whitespace && go install github.com/ultraware/whitespace/cmd/whitespace@latest
    ! is_bin_in_path wsl && go install github.com/bombsimon/wsl/v4/cmd/wsl@latest
    ! is_bin_in_path mirror && go install github.com/butuzov/mirror/cmd/mirror@latest
    ! is_bin_in_path misspell && go install github.com/client9/misspell/cmd/misspell@latest
    ! is_bin_in_path perfsprint && go install -v github.com/catenacyber/perfsprint@latest
    ! is_bin_in_path testifylint && go install github.com/Antonboom/testifylint@latest
    ! is_bin_in_path usestdlibvars && go install github.com/sashamelentyev/usestdlibvars@latest
    ! is_bin_in_path gci && GOBIN=$HOME/khan/webapp/genfiles/go/bin go install github.com/daixiang0/gci@latest
    ! is_bin_in_path golines && GOBIN=$HOME/khan/webapp/genfiles/go/bin go install github.com/segmentio/golines@latest
    ! is_bin_in_path gofumpt && GOBIN=$HOME/khan/webapp/genfiles/go/bin go install mvdan.cc/gofumpt@latest
    ! is_bin_in_path golangci-lint && GOBIN=$HOME/go/bin curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $GOBIN latest
    for D in */; do
        if [ -f "${D}go.mod" ]; then
            echo -e "\033[32m\xE2\x9c\x93 Running linting for ${D}\n\n\033[0m"
            cd "${D}"
            whitespace --fix ./... 2>&1 > /dev/null
            mirror --fix ./... 2>&1 > /dev/null
            perfsprint --fix ./... 2>&1 > /dev/null
            testifylint --suite-extra-assert-call.mode=require --fix ./... 2>&1 > /dev/null
            # usestdlibvars --fix ./... 2>&1 > /dev/null # noisy on generated files
            gci write -s standard -s default -s "prefix(github.com/Khan)" --skip-generated . 2>&1 > /dev/null
            gofumpt -w . 2>&1 > /dev/null
            wsl --fix . 
            "${HOME}/go/bin/golangci-lint" run --fix --config=~/Documents/git/districts-jobs/.golangci.yml ./...
            cd ..
        fi
    done
}

lintall
```

While some `govet` linters supports SuggestedFixes, `go vet` does not have a `--fix` flag to apply them. One of it's linters provides an independent entrypoint that does:
```
go install golang.org/x/tools/go/analysis/passes/fieldalignment/cmd/fieldalignment
fieldalignment -fix ./...
```
The other `govet` linters that support SuggestedFixes without an obvious corresponding installable entrypoint that can apply them: 
+ assign
+ composite
+ sigchanyzer
+ sortslice
+ stringintconv
+ timeformat
+ unreachable

**Comment by ldez** [comment]
I will handle this issue.

**Comment by ldez** [comment]
**UPDATE**

I will give some details of my current about the topic.

My plan is to replace the current implementation (based on `Replacement`, `HunkPos`, etc.) with an implementation based on `analysis.SuggestedFix`.
But this is very complex because the current implementation "leaks" everywhere, and some linter implementations (`gofmt`, `goimport`, etc.) are heavily linked to this implementation.

FYI, the current implementation is based on patches (`diff` style), and this code is not "friendly", and is completely different than something based `analysis.SuggestedFix`.
It requires rewriting all the implementation from scratch.

It's a long job, so it will take me more than one week.

Maybe at the end, I will just create a kind of "compatibility layer".
This solution will be a type of failure to me because it will increase complexity instead of decreasing it, and so the goal of simplifying the maintenance will not be achieved.

**Comment by ldez** [comment]
**Update**

I have a working PoC (it's a draft), I keep it locally for now because it's a bit too raw.

Currently, golangci-lint is able to "autofix" reports from 13 linters:
- `gci`
- `gocritic`
- `godot`
- `gofmt`
- `gofumpt`
- `goheader`
- `goimports`
- `mirror`
- `misspell`
- `nolintlint`
- `protogetter`
- `tagalign`
- `whitespace`

I converted 8 linters of those currently supported linters to `SuggestedFix`.
So, 5 of the 13 are still not working (`gofmt`, `gofumpt`, `goheader`, `goimports`, `nolintlint`)

About the other linters, I'm able to use "autofix" for the totality of linters that support `SuggestedFix`:

- `canonicalheader`
- `dupword`
- `fatcontext`
- `ginkgolinter`
- `err113`
- `errorlint`
- `iface`
- `importas`
- `intrange`
- `nakedret`
- `nlreturn`
- `perfsprint`
- `testifylint`
- `wsl`
- `govet`

Also, the "autofix" system is less powerful: there are some limitations on fixes on the same line, for now, I don't know if I will be able to improve that.

This is still under development, and until it is fully working. For now, I do not consider my current implementation as a working solution.

**Comment by JoelSpeed** [comment]
> Also, the "autofix" system is less powerful: there are some limitations on multiple fixes on the same line.

Could you expand on this point? My understanding of SuggestedFixes was that it could apply fixes as long as they didn't overlap, so you have to specify starts and ends if you have multiple fixes on the same line. Was that not true of the previous implementation? How did it handle overlapping fixes?

**Comment by ldez** [comment]
My sentence was not clear: the new system is less powerful.

Remember, this is a rewrite from scratch. https://github.com/golangci/golangci-lint/issues/1779#issuecomment-2526377728

**Comment by bombsimon** [comment]
> Was that not true of the previous implementation? How did it handle overlapping fixes?

The previous (current) implementation does not support this, e.g #1490, #4885, #1510, #3819, 3454.

I don't know what the new implementation looks like but the idea of using `SuggestedFixes` sounds like it would use the data from `pass.Report`/`pass.Diagnostic` which holds the `SuggestedFixes` slice. Given this it might be possible at some point to apply overlapping fixes or at least fixes on the same line with different pos if they share the pass.

I think for now, just using `SuggestedFixes` is a huge thing and probably not worth focusing on overlapping fixes just yet.

**Comment by ldez** [comment]
> The previous (current) implementation does not support this

This is more futile: the current implementation is able in some cases support that.

For example, run `golangci-lint run --enable-only misspell --fix` on this:
```go
package sandbox

import "log"

// langauge lala
// lala langauge
// langauge
// langauge langauge

// check Langauge
// and check langAuge

func langaugeMisspell() {
	var langauge, langaugeAnd string
	log.Printf("it's becouse of them: %s, %s", langauge, langaugeAnd)
}

```

This example contains several problems on the same line.

**Comment by pohly** [comment]
Perhaps overlapping fixes can be handled incrementally, as in "apply some fixes that don't overlap, rerun analysis, apply again, until nothing pops up anymore"? Just a thought.

**Comment by ldez** [comment]
**Update**

I think I win the fight against the most complex part: the "formaters" (`gofmt`, `gofumpt`, `goimports`).
This led me to deeper thoughts about how "formaters" are integrated, but that's another topic.

So the 12 currently supported linters, and 15 linters that support `SuggestedFix` are working:
- `canonicalheader` (new)
- `dupword` (new)
- `err113` (new)
- `errorlint` (new)
- `fatcontext` (new)
- `gci`
- `ginkgolinter` (new)
- `gocritic`
- `godot`
- `gofmt`
- `gofumpt`
- `goheader`
- `goimports`
- `govet` (new)
- `iface` (new)
- `importas` (new)
- `intrange` (new)
- `mirror`
- `misspell`
- `nakedret` (new)
- `nlreturn` (new)
- `perfsprint` (new)
- `protogetter`
- `tagalign`
- `testifylint` (new)
- `whitespace`
- `wsl` (new)

Now, I have to handle `nolintlint`.

Reports on the same line (not overlapping) are still not supported.
There are several other problems that I should solve too.

This is still in WIP.
Until it is fully working, I do not consider my current implementation as a working solution.

If you want to help me:
- https://bento.me/golangci-lint
- https://bento.me/ldez

**Comment by ldez** [comment]
**Update**

The problem of changes on the same line is fixed: the fun fact is it was just a missing condition inside an `if`.

---

Also overlapping is now managed in a stable way.

The situation:

1. You have 2 linters (gofmt and goimports for example)
2. The 2 linters produce 2 reports/changes on the same line (ex: an import on L4)) -> overlap.
3. The 2 same linters have also produced reports/changes but without overlap (ex: changes an import on L5 and another on L6)

This situation is 99% related to formaters because they want to apply changes to the same sections but not in the same way, so not in the same position.

Before/Currently:

The overlap was managed by picking "randomly" one change for one of both linters (L4).
And all the other changes (without overlap) from the 2 linters were applied (L5, L6).

The result was a mix of non-consistent changes: deletion of imports, duplication of imports, etc.

The problem here is not the overlaps by themselves (because they are "managed") but the other changes.

With my implementation:

If there are overlaps, only one of the 2 linters is "allowed" to apply changes on the targeted file (ex: L4, L5, not L6).
The other linters that don't overlap are also "allowed" to apply changes to this file.

I tested my implementation with the examples from the issues, and it works as expected: no unexpected deletions, and no duplications.

---

I still have to handle `nolintlint`, it is not the easiest to migrate.
(maybe my next work will be to deprecate it and implement a new directive system, but this is another topic)

I'm increasingly confident in my current implementation.
Now, IMHO, this is a working solution.

---

If you benefit from this change, and you want to help me:

- https://bento.me/golangci-lint
- https://bento.me/ldez

**Comment by ldez** [comment]
**Update**

`nolintlint` was a pain to migrate but this is done.

The "autofix" is now working for 28 linters.

Now, I should clean the implementation, add new linter tests, create a reviewable PR, and maybe add some bonus.

It will take one or two days so I will create a PR this weekend.

The go1.24.0-rc1 release will happen very soon, I already prepared a PR but this could take some time too.

This will be the last update in this issue as the next step is to create a PR.

---

I already worked 10 days on this topic, and my only revenue is donations.
During this period golangci-lint received $5 ( :heart: ) and I received $0.
It's a bit sad, but not surprising, I need to think about how to improve that.

If you enjoy this change, and you or your company want to help me:

- https://bento.me/golangci-lint
- https://bento.me/ldez

**Comment by mitar** [comment]
We use thanks.dev for donation distribution and I just added https://github.com/golangci/golangci-lint as a manual inclusion. You can quickly sign up and then it connects with your GitHub sponsors, if that would work for you. Thank you for your great work!

**Comment by StevenACoffman** [comment]
It's not much I know, but I just put in $20 for you for this, and am still waiting to hear if my employer will sponsor ongoing.

**Comment by ldez** [comment]
**Update**

Sadly, I have an "extra" update :cry: 

I'm fighting against CGO directives and friends, so I must add tests everywhere and adjust the linter implementations when needed.
This takes a lot of time.

I do my best but I will not be able to open a PR today.

**Comment by Crocmagnon** [comment]
Just threw 20€ in the hat, thank you for your continued work!
Let me know if anything is needed on `fatcontext`.

**Comment by ldez** [comment]
**Update**

After a very very long tunnel to add tests, run tests, and analyze the results, I have some conclusions:
- yacc directives have never worked for 95% of the linters
- cgo is not supported by 13 linters. With my changes (and some PR I opened on linters) I fixed 2 linters but broke 3 other linters.

I evaluate this topic as independent: I postpone the problems with cgo and yacc, I will handle this later.

I will remove all the yacc tests (but keep the cgo tests), clean everything, and open a PR.

It was a difficult and exciting tunnel, a bit extreme, and now I need some rest.
